File: flutter_inappwebview_windows/windows/types/client_cert_challenge.cpp
#include "../utils/flutter.h"
#include "client_cert_challenge.h"

namespace flutter_inappwebview_plugin
{
  ClientCertChallenge::ClientCertChallenge(const std::shared_ptr<URLProtectionSpace> protectionSpace,
    const std::vector<std::string>& allowedCertificateAuthorities,
    const bool& isProxy,
    const std::vector<std::shared_ptr<SslCertificate>>& mutuallyTrustedCertificates)
    : URLAuthenticationChallenge(protectionSpace), allowedCertificateAuthorities(allowedCertificateAuthorities), isProxy(isProxy), mutuallyTrustedCertificates(mutuallyTrustedCertificates)
  {}

  flutter::EncodableMap ClientCertChallenge::toEncodableMap() const
  {
    auto map = URLAuthenticationChallenge::toEncodableMap();
    map.insert({ "allowedCertificateAuthorities", make_fl_value(allowedCertificateAuthorities) });
    map.insert({ "isProxy", make_fl_value(isProxy) });
    map.insert({ "mutuallyTrustedCertificates", make_fl_value(functional_map(mutuallyTrustedCertificates, [](const std::shared_ptr<SslCertificate>& item) { return item->toEncodableMap(); })) });
    return map;
  }
}

File: flutter_inappwebview_platform_interface/lib/src/types/client_cert_challenge.dart
import 'package:flutter_inappwebview_internal_annotations/flutter_inappwebview_internal_annotations.dart';

import 'url_authentication_challenge.dart';
import 'url_protection_space.dart';
import '../in_app_webview/platform_webview.dart';
import '../types/ssl_certificate.dart';
import 'enum_method.dart';

part 'client_cert_challenge.g.dart';

///Class that represents the challenge of the [PlatformWebViewCreationParams.onReceivedClientCertRequest] event.
///It provides all the information about the challenge.
@ExchangeableObject()
class ClientCertChallenge_ extends URLAuthenticationChallenge_ {
  ///Use [principals] instead.
  @Deprecated('Use principals instead')
  List<String>? androidPrincipals;

  ///The acceptable certificate issuers for the certificate matching the private key.
  @SupportedPlatforms(
    platforms: [
      AndroidPlatform(
        apiName: "ClientCertRequest.getPrincipals",
        apiUrl:
            "https://developer.android.com/reference/android/webkit/ClientCertRequest#getPrincipals()",
        available: "21",
      ),
    ],
  )
  List<String>? principals;

  ///Use [keyTypes] instead.
  @Deprecated('Use keyTypes instead')
  List<String>? androidKeyTypes;

  ///Returns the acceptable types of asymmetric keys.
  @SupportedPlatforms(
    platforms: [
      AndroidPlatform(
        apiName: "ClientCertRequest.getKeyTypes",
        apiUrl:
            "https://developer.android.com/reference/android/webkit/ClientCertRequest#getKeyTypes()",
        available: "21",
      ),
    ],
  )
  List<String>? keyTypes;

  ///The collection contains Base64 encoding of DER encoded distinguished names
  ///of certificate authorities allowed by the server.
  @SupportedPlatforms(platforms: [WindowsPlatform()])
  List<String>? allowedCertificateAuthorities;

  ///If the server that issued this request is an http proxy.
  @SupportedPlatforms(platforms: [WindowsPlatform()])
  bool? isProxy;

  ///The collection contains mutually trusted CA certificates.
  @SupportedPlatforms(platforms: [WindowsPlatform()])
  List<SslCertificate_>? mutuallyTrustedCertificates;

  ClientCertChallenge_({
    required URLProtectionSpace_ protectionSpace,
    @Deprecated('Use principals instead') this.androidPrincipals,
    this.principals,
    @Deprecated('Use keyTypes instead') this.androidKeyTypes,
    this.keyTypes,
    this.allowedCertificateAuthorities,
    this.isProxy,
    this.mutuallyTrustedCertificates,
  }) : super(protectionSpace: protectionSpace);
}
